pulser/waveforms.py

```diff
@@ -373,7 +463,7 @@ def samples(self) -> np.ndarray:
         Returns:
             numpy.ndarray: A numpy array with a value for each time step.
         """
-        return np.linspace(self._start, self._stop, num=self._duration)
+        return self._samples
```

Comment:
The problem with doing this is (and the reason it wasn't done) that it give direct access to the samples, so the user could change it in-place:
```
samps = wf.samples
samps = np.array([1, 2 , 3])
```
After this block, `wf.samples` will be `np.array([1, 2, 3])`
I can't figure out a way out of it without doing the array from scratch or returning a `copy`. A compromise is to have `self.samples` return `self._samples.copy()` but then use `self._samples` internally.

Reply:
> The problem with doing this is (and the reason it wasn't done) that it give direct access to the samples, so the user could change it in-place:
> 
> ```
> samps = wf.samples
> samps = np.array([1, 2 , 3])
> ```
> 
> After this block, `wf.samples` will be `np.array([1, 2, 3])`
> I can't figure out a way out of it without doing the array from scratch or returning a `copy`. A compromise is to have `self.samples` return `self._samples.copy()` but then use `self._samples` internally.

Just to make sure: that applies to CustomWaveform and Rampwaveform that both return `self._samples`?

Reply:
You're right, `CustomWaveform` is returning `self._samples` although perhaps it shouldn't be. But `RampWaveform` is currently not returning `self._samples`, it's this change that would make it do so.